### PR TITLE
Add chart for Traefik-based ingress controller

### PR DESCRIPTION
**Pull request by krancour**
Current Todo:

- [ ] Decide on whether this chart should be in stable or incubator

### CODE REVIEW

**Comment by prydonius** [comment]
Thanks @krancour!

File: incubator/traefik/README.md
Comment (prydonius) [inline incubator/traefik/README.md]: Is this a requirement? I think we generally prefer that Helm manages the namespace Charts get installed into with the `--namespace` flag.

Comment (krancour) [reply incubator/traefik/README.md]: It's not a hard and fast requirement, however, since ingress controllers are frequently used to watch `Ingress` objects _cluster-wide_ (i.e. in all namespaces), they often require the elevated permissions of the `kube-system` namespace's default service account.

When writing this chart, I assumed that for the less-frequent (or so I imagine) use case where someone _does_ want to install this into some other namespace that `--namespace` would _override_ the hard-coded default of `kube-system`. I've put that to the test now, however, and found it doesn't work that way!

cc @technosophos @michelleN isn't `--namespace` supposed to override whatever default is hard-coded in the chart's manifests? Or was it just wishful thinking on my part that it works that way?

At any rate, it does seem that what is best for now is that I omit this hard-coded default and instead _document_ that this component would most often be installed into `kube-system`.

Comment (krancour) [reply incubator/traefik/README.md]: @prydonius I've amended the PR as discussed above.

File: incubator/traefik/LICENSE
Comment (chrislovecnm) [inline incubator/traefik/LICENSE]: Do we need this? This project is Apache 2 licensed and now we have MIT?

File: incubator/traefik/templates/deployment.yaml
Comment (chrislovecnm) [inline incubator/traefik/templates/deployment.yaml]: ready and lively probes please, as well as resource limits ;)

File: incubator/traefik/templates/deployment.yaml
Comment (chrislovecnm) [inline incubator/traefik/templates/deployment.yaml]: It there value to have storage classes?

Comment (chrislovecnm) [reply incubator/traefik/templates/deployment.yaml]: It == Is

**Review by chrislovecnm — CHANGES_REQUESTED** [review]
Thanks for the PR, looks awesome. Have some recommended tweaks

File: incubator/traefik/LICENSE
Comment (krancour) [reply incubator/traefik/LICENSE]: Well, that's a tricky question. It seems individual charts are subject to their own license, no? What I opted to do here was use the same license that Traefik itself uses. If that's the wrong move, lmk.

File: incubator/traefik/templates/deployment.yaml
Comment (krancour) [reply incubator/traefik/templates/deployment.yaml]: Yep. Good call.

File: incubator/traefik/templates/deployment.yaml
Comment (krancour) [reply incubator/traefik/templates/deployment.yaml]: I'm not sure what you're asking here. Can you clarify please?

This currently follows the exact pattern used by the stable MySQL chart for persistent storage.

File: incubator/traefik/LICENSE
Comment (viglesiasce) [reply incubator/traefik/LICENSE]: I think this code/repo can have a separate license from the source of traefik itself.

File: incubator/traefik/values.yaml
Comment (viglesiasce) [inline incubator/traefik/values.yaml]: Can this be a secret name instead? Then the docs could outline how to create that secret (keys/values) such that it can be used here?

File: incubator/traefik/templates/deployment.yaml
Comment (viglesiasce) [reply incubator/traefik/templates/deployment.yaml]: The storage class is defined in the PVC not the deployment. Nothing to see here.

**Comment by viglesiasce** [comment]
Awesome stuff @krancour!

**Comment by viglesiasce** [comment]
@k8s-bot ok to test

File: incubator/traefik/LICENSE
Comment (krancour) [reply incubator/traefik/LICENSE]: Ok. I'll change it.

File: incubator/traefik/values.yaml
Comment (krancour) [reply incubator/traefik/values.yaml]: If I can make an argument for keeping it as it is...

The base64 encoded cert and key you put here _becomes_ a secret. This means that you provide customizations to this either through `--set` on the command line or in your own `values.yaml`. i.e. You can customize these through the usual Helm workflow for customizing a chart.

Compare that to some other approach that requires a careful reading of the documentation and an "out of band" effort to create the secrets... I don't believe that compares favorably to what's in place now.

Comment (viglesiasce) [reply incubator/traefik/values.yaml]: Fair. I'm fine as is. If users want something different, I'm sure they will chime in down the road.

Could I ask you to add instructions on getting those values? Also what do the default certs point to?

Comment (krancour) [reply incubator/traefik/values.yaml]: The readme has a table describing the configuration options. It's clear that the cert and key should be base64 encoded. I can add more specific instructions if you like. lmk

The cert that's included is a self-signed cert for *.example.com-- so it's basically garbage. The benefit, however, is that it's enough to get people going.

I will update the table to reflect the details of this cert.

**Comment by ekozan** [comment]
@krancour  Can you integrate HostPort option or I'll make a PR after merge

File: incubator/traefik/README.md
Comment (ekozan) [inline incubator/traefik/README.md]: Actually it's incubator/traefik not stable/traefik

**Comment by krancour** [comment]
@ekozan, did you mean `NodePort`? You want that as an alternative to `type: LoadBalancer`?

That seems a reasonable option (but not a default). It would certainly be a benefit to anyone who wants to run this anywhere where the "cloud provider" no-ops provisioning of an external load balancer-- e.g. minikube or bare metal.

I'd be happy to add that.

File: incubator/traefik/README.md
Comment (vdice) [inline incubator/traefik/README.md]: a couple minor typos here; perhaps the following were intended: `... direct inbound traffic to the load balancer` and `You can use the command ...`

Comment (krancour) [reply incubator/traefik/README.md]: Nice eye, @vdice. Will fix.

**Comment by viglesiasce** [comment]
Tested manually and all things seem to check out. Thanks again @krancour! Let me know when you've patched the README fixes. Also if you'd like to get this into stable, adding a NOTES.txt pointing people in the right direction for how to configure and use traefik would do the trick.

**Comment by krancour** [comment]
@viglesiasce great. I will try and amend with all suggested fixes as soon as time permits-- a few hours off, for sure.

**Comment by ekozan** [comment]
Yeap it's an alternative for barebone cluster ;)

We need 
Add option for switch the default type of service: LoadBalancer / HostPort / NodePort  
Add option for switch to Deamonset is needed too :p i'll handle this if you dont want to add this one

**Comment by ultimateboy** [comment]
Just thought I'd report my successful testing of this chart, and Traefik's ability to issue certs from Let's Encrypt.

```
helm install incubator/traefik --name my-traefik --namespace kube-system --set dashboard.enabled=true,dashboard.domain=traefik.domain.you.own.com,ssl.enabled=true,acme.enabled=true,acme.email=admin@domain.you.own.com
```

```
kubectl run nginx --image=nginx
kubecl expose deploy nginx --port=80 --target-port=80
```

```
echo "apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: nginx
  namespace: default
spec:
  rules:
  - host: some.domain.you.own.com
    http:
      paths:
      - backend:
          serviceName: nginx
          servicePort: 80" > ingress.yaml

kubectl create -f ingress.yaml
```

```
echo | openssl s_client -servername some.domain.you.own.com -connect some.domain.you.own.com:443 2>/dev/null | openssl x509 -noout -issuer
issuer= /C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
```

**Comment by krancour** [comment]
It'll be tomorrow before I get to this.

File: incubator/traefik/README.md
Comment (chrislovecnm) [reply incubator/traefik/README.md]: If he adds a NOTES.txt it can be merged into stable directly

File: incubator/traefik/LICENSE
Comment (krancour) [reply incubator/traefik/LICENSE]: I'm actually just going to remove the license from this chart altogether. It seems license is optional for charts and no other chart in this repo includes one. I'm going to follow that precedent.

**Comment by krancour** [comment]
@viglesiasce I've addressed comments apart from one. You indicated I should add a `NOTES.txt` "pointing people in the right direction for how to configure and use traefik."

I'm unclear on what that should contain or what its purpose would be.

`charts.yaml` already points users in the direction of the Traefik documentation. That aside, detailed knowledge of configuring Traefik isn't required to use this chart, since all customizations are carried out not by editing Traefik's config directly, but by supplying a `values.yaml` instead.

**Comment by krancour** [comment]
@ultimateboy offline you asked me to support LE staging to avoid rate limits when testing / experimenting. I've added that now and made it the default as well (when ACME is enabled). Check it out and lmk what you think.

**Comment by ultimateboy** [comment]
@krancour Awesome! It's probably worth explicitly calling that configuration out in the README as people might be confused why they aren't getting real valid certs. But this looks great!

**Comment by krancour** [comment]
@viglesiasce nevermind re: my previous question about notes.txt... I realize now that this is an optional, but conventional component of charts. I've looked at some existing charts to get a sense of what sort of info is usually included. Pending the addition of those notes, I think this PR is into a reviewable state.

**Comment by chrislovecnm** [comment]
Before we merge you mind squashing your commits into units of work? Thanks

**Comment by krancour** [comment]
@chrislovecnm I was going to squash everything into one commit when done. Is that acceptable?

**Comment by chrislovecnm** [comment]
@krancour your call mate!

**Comment by viglesiasce** [comment]
No need to squash on your end as Github now allows Squash and merge when accepting the pull request.

**Comment by krancour** [comment]
Ya... knew that, but not all projects necessarily have that enabled. If it's enabled here, awesome.

**Comment by lachie83** [comment]
I've reached out to @krancour and have agreed to work on getting a NOTES.txt into this chart.

**Comment by lachie83** [comment]
Seeing the following in the log file

```
time="2016-11-02T05:00:14Z" level=error msg="Kubernetes connection error failed to decode watch event: GET \"https://10.3.0.1:443/apis/extensions/v1beta1/ingresses\" : unexpected EOF, retrying in 682.444463ms" 
```

Looks like it's an issue with traefik -- https://github.com/containous/traefik/issues/732

**Comment by lachie83** [comment]
Tried putting an ingress in-front of Jenkins and it doesn't look like it handles more complex backends. Found this open issue -- https://github.com/containous/traefik/issues/30

**Comment by lachie83** [comment]
Added NOTES.txt. I believe this is the only outstanding item on this PR. Please review

**Comment by lachie83** [comment]
@krancour Thanks for updating NOTES.txt. Looks good!

**Comment by lachie83** [comment]
@viglesiasce @prydonius @krancour I would like to propose that we reopen the discussion on moving this to the incubator. Based on my testing last night I found several things that didn't work and associated open issues on the Traefik repo. By putting this chart into stable may give the impression that it's ready for primetime. 

To sum up my thoughts here. The chart is stable, the application is still maturing.

File: stable/traefik/README.md
Comment (lachie83) [inline stable/traefik/README.md:12]: Can you a disclaimer about the state of Traefik?

File: stable/traefik/README.md
Comment (lachie83) [inline stable/traefik/README.md:11]: Looks great! Thanks for the updates!

**Review by lachie83 — CHANGES_REQUESTED** [review]
Can we add a disclaimer on the current state of Traefik and it's prod readiness?

**Review by lachie83 — COMMENTED** [review]
NOTES.txt looks much better! Thanks

File: stable/traefik/README.md
Comment (krancour) [reply stable/traefik/README.md:12]: Yep. Meant to, as we'd discussed. Forgot. Coming atcha.

**Comment by prydonius** [comment]
@lachie83 @viglesiasce @krancour 

> By putting this chart into stable may give the impression that it's ready for primetime.

Agreed, if the project itself is not production-ready then the package should be in incubator.

File: stable/traefik/README.md
Comment (lachie83) [inline stable/traefik/README.md:6]: Thanks @krancour

**Review by lachie83 — APPROVED** [review]

**Comment by krancour** [comment]
To clarify, I believe Traefik regards itself as production ready and people certainly _do_ use Traefik in production. Any assertion about production readiness here is purely based on @lachie83's and my own hesitation to endorse this as such.

**Comment by krancour** [comment]
So I'm going to move this to incubator. Standby.

**Comment by krancour** [comment]
Moved

**Comment by lachie83** [comment]
> To clarify, I believe Traefik regards itself as production ready and people certainly do use Traefik in production. Any assertion about production readiness here is purely based on @lachie83's and my own hesitation to endorse this as such.

Agreed 👍

**Comment by viglesiasce** [comment]
I dont know about that folks. Where is this production ready requirement coming from for stable? Its not listed anywhere and the other requirements are quite explicit. I don't want us to cross this line.

**Comment by krancour** [comment]
So if the chart's own author(s) would hesitate to use it in production (depending on the use case), this can still qualify as "stable?"

**Comment by krancour** [comment]
Maybe some clarification on what, precisely, is meant by "stable" is in order. I will acknowledge I made a pretty big (and perhaps incorrect) assumption that stability was equated with production readiness.

**Comment by lachie83** [comment]
> I dont know about that folks. Where is this production ready requirement coming from for stable? Its not listed anywhere and the other requirements are quite explicit. I don't want us to cross this line.

Understood and I agree that it's not our position to decide whether something is production ready. I don't want to cross that line. Stable does not mean Prod Ready

How about we approach this from a difference angle - putting this in the incubator is a great way for it to gather some miles

**Comment by viglesiasce** [comment]
Stable charts meet [these requirements](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#how-to-contribute-a-chart).

If we cross the line of requiring things to be production ready then we need to define production ready in a way that is obvious otherwise we are going to end up with a lot of bikeshedding. For most packages (rpm and deb) the defaults are sane but not production ready.

**Comment by krancour** [comment]
> How about we approach this from a difference angle - putting this in the incubator is a great way for it to gather some miles

@viglesiasce is there any validity to that notion? That things are "promoted" so stable after proving themselves in the incubator?

**Comment by viglesiasce** [comment]
The only thing we say in the [README.md](https://github.com/kubernetes/charts#repository-structure) is that stable charts meet the criteria in the [technical requirements](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#how-to-contribute-a-chart). If we want to change that policy then we need to update the documentation via a PR where we can hash out the why and how.

**Comment by krancour** [comment]
Ok. I'm going to strike the last commit (the move to incubator) from this PR.

**Comment by lachie83** [comment]
Thanks for the great discussion and clarification. 

Given that this chart meets the requirements. My vote is on putting this in stable.

**Comment by krancour** [comment]
All set. Back in stable.

**Comment by prydonius** [comment]
@viglesiasce you're right, for some reason I thought some level of 'production-readiness' was a requirement. I would argue for promoting the stable repo as a source of 'production-ready' charts, but it's a separate discussion to define what we could mean by that and whether it makes sense to enforce it.

**Comment by lachie83** [comment]
LGTM

**Comment by viglesiasce** [comment]
LGTM

**Comment by prydonius** [comment]
@krancour lgtm, all good to merge?

**Comment by krancour** [comment]
Yes

**Comment by lachie83** [comment]
Thanks all! Excited to have this chart!

**Pull request merged**
